File: backend/src/messages.rs
use bytes::{Buf, Bytes};
use miette::{Context, IntoDiagnostic, bail};
use serde_bytes::ByteBuf;
use uuid::Uuid;

use crate::Epoch;

#[derive(Clone, Debug)]
pub enum NGMessage {
    StateSubmitting(Epoch),
    SubmitName(String),
    NameSubmitted(String, Uuid),
    UnsubmitName(Uuid),
    NameUnsubmitted(Uuid),
    NumNames(usize),
    RequestPlayingState,
    Names(Vec<String>, Vec<u8>),
    GuessName(usize),
    NameGuessed(usize),
    UnguessName(usize),
    NameUnguessed(usize),
    RequestSubmittingState,
}

impl NGMessage {
    pub fn parse(mut bytes: Bytes) -> miette::Result<Self> {
        let typ = bytes.get_u32();
        match typ {
            0 => Ok(NGMessage::StateSubmitting(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from StateSubmitting message")?,
            )),
            1 => Ok(NGMessage::SubmitName(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from SubmitName message")?,
            )),
            2 => {
                let (name, id) = rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from NameSubmitted message")?;
                Ok(NGMessage::NameSubmitted(name, id))
            }
            3 => Ok(NGMessage::UnsubmitName(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from UnsubmitName message")?,
            )),
            4 => Ok(NGMessage::NameUnsubmitted(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from NameUnsubmitted message")?,
            )),
            5 => Ok(NGMessage::NumNames(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from NumNames message")?,
            )),
            6 => {
                if bytes.len() != 0 {
                    bail!(
                        "nonzero length in RequestPlayingState message: {}",
                        bytes.len()
                    );
                } else {
                    Ok(NGMessage::RequestPlayingState)
                }
            }
            7 => {
                let (names, guesses): (Vec<String>, ByteBuf) = rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from Names message")?;
                Ok(NGMessage::Names(names, guesses.into_vec()))
            }
            8 => Ok(NGMessage::GuessName(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from GuessName message")?,
            )),
            9 => Ok(NGMessage::NameGuessed(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from NameGuessed message")?,
            )),
            10 => Ok(NGMessage::UnguessName(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from UnguessName message")?,
            )),
            11 => Ok(NGMessage::NameUnguessed(
                rmp_serde::from_slice(&bytes)
                    .into_diagnostic()
                    .wrap_err("parse content from NameUnguessed message")?,
            )),
            12 => {
                if bytes.len() != 0 {
                    bail!(
                        "nonzero length in RequestSubmittingState message: {}",
                        bytes.len()
                    );
                } else {
                    Ok(NGMessage::RequestSubmittingState)
                }
            }
            _ => {
                bail!("message has unknown type: {typ}");
            }
        }
    }

    pub fn encode(&self) -> Bytes {
        let mut encoded = vec![0; 4];

        encoded[..4].copy_from_slice(
            &match self {
                NGMessage::StateSubmitting(_) => 0u32,
                NGMessage::SubmitName(_) => 1,
                NGMessage::NameSubmitted(_, _) => 2,
                NGMessage::UnsubmitName(_) => 3,
                NGMessage::NameUnsubmitted(_) => 4,
                NGMessage::NumNames(_) => 5,
                NGMessage::RequestPlayingState => 6,
                NGMessage::Names(_, _) => 7,
                NGMessage::GuessName(_) => 8,
                NGMessage::NameGuessed(_) => 9,
                NGMessage::UnguessName(_) => 10,
                NGMessage::NameUnguessed(_) => 11,
                NGMessage::RequestSubmittingState => 12,
            }
            .to_be_bytes(),
        );

        match self {
            NGMessage::StateSubmitting(epoch) => {
                rmp_serde::encode::write(&mut encoded, epoch).unwrap()
            }
            NGMessage::SubmitName(name) => rmp_serde::encode::write(&mut encoded, name).unwrap(),
            NGMessage::NameSubmitted(name, id) => {
                rmp_serde::encode::write(&mut encoded, &(name, id)).unwrap()
            }
            NGMessage::UnsubmitName(id) => rmp_serde::encode::write(&mut encoded, id).unwrap(),
            NGMessage::NameUnsubmitted(id) => rmp_serde::encode::write(&mut encoded, id).unwrap(),
            NGMessage::NumNames(num) => rmp_serde::encode::write(&mut encoded, num).unwrap(),
            NGMessage::RequestPlayingState => {}
            NGMessage::Names(names, guesses) => {
                rmp_serde::encode::write(&mut encoded, &(names, serde_bytes::Bytes::new(&guesses)))
                    .unwrap()
            }
            NGMessage::GuessName(index) => rmp_serde::encode::write(&mut encoded, index).unwrap(),
            NGMessage::NameGuessed(index) => rmp_serde::encode::write(&mut encoded, index).unwrap(),
            NGMessage::UnguessName(index) => rmp_serde::encode::write(&mut encoded, index).unwrap(),
            NGMessage::NameUnguessed(index) => {
                rmp_serde::encode::write(&mut encoded, index).unwrap()
            }
            NGMessage::RequestSubmittingState => {}
        }

        Bytes::from(encoded)
    }
}

File: frontend/src/lib/messages.ts
import { decode, encode } from '@msgpack/msgpack';
import { parse, stringify } from 'uuid';

export enum MessageType {
  StateSubmitting,
  SubmitName,
  NameSubmitted,
  UnsubmitName,
  NameUnsubmitted,
  NumNames,
  RequestPlayingState,
  Names,
  GuessName,
  NameGuessed,
  UnguessName,
  NameUnguessed,
  RequestSubmittingState,
}

export type Uuid = string;

export type StateSubmittingMessage = {
  type: MessageType.StateSubmitting;
  content: number;
};

export type SubmitNameMessage = {
  type: MessageType.SubmitName;
  content: string;
};

export type NameSubmittedMessage = {
  type: MessageType.NameSubmitted;
  content: [string, Uuid];
};

export type UnsubmitNameMessage = {
  type: MessageType.UnsubmitName;
  content: Uuid;
};

export type NameUnsubmittedMessage = {
  type: MessageType.NameUnsubmitted;
  content: Uuid;
};

export type NumNamesMessage = {
  type: MessageType.NumNames;
  content: number;
};

export type RequestPlayingStateMessage = {
  type: MessageType.RequestPlayingState;
  content: null;
};

export type NamesMessage = {
  type: MessageType.Names;
  content: [string[], boolean[]];
};

export type GuessNameMessage = {
  type: MessageType.GuessName;
  content: number;
};

export type NameGuessedMessage = {
  type: MessageType.NameGuessed;
  content: number;
};

export type UnguessNameMessage = {
  type: MessageType.UnguessName;
  content: number;
};

export type NameUnguessedMessage = {
  type: MessageType.NameUnguessed;
  content: number;
};

export type RequestSubmittingStateMessage = {
  type: MessageType.RequestSubmittingState;
  content: null;
};

export type Message =
  | StateSubmittingMessage
  | SubmitNameMessage
  | NameSubmittedMessage
  | UnsubmitNameMessage
  | NameUnsubmittedMessage
  | NumNamesMessage
  | RequestPlayingStateMessage
  | NamesMessage
  | GuessNameMessage
  | NameGuessedMessage
  | UnguessNameMessage
  | NameUnguessedMessage
  | RequestSubmittingStateMessage;

function bitfieldToBooleanArray(
  bitfield: Uint8Array,
  arrayLength: number,
): boolean[] {
  const numBytesToProcess = Math.min(
    bitfield.length,
    Math.ceil(arrayLength / 8),
  );
  const array: boolean[] = new Array(numBytesToProcess * 8);

  for (let i = 0; i < numBytesToProcess; i++) {
    const num = bitfield[i];
    const base = i * 8;
    array[base] = !!(num & 0x80);
    array[base + 1] = !!(num & 0x40);
    array[base + 2] = !!(num & 0x20);
    array[base + 3] = !!(num & 0x10);
    array[base + 4] = !!(num & 0x8);
    array[base + 5] = !!(num & 0x4);
    array[base + 6] = !!(num & 0x2);
    array[base + 7] = !!(num & 0x1);
  }

  if (array.length < arrayLength) {
    array.push(...new Array(arrayLength - array.length).fill(false));
  } else {
    array.splice(arrayLength);
  }

  return array;
}

export function decodeMessage(message: ArrayBuffer): Message {
  const view = new DataView(message);
  const type = view.getUint32(0);
  if (view.byteLength === 4) {
    return { type, content: null };
  }

  let content = decode(new Uint8Array(message, 4)) as Message['content'];
  switch (type) {
    case MessageType.Names: {
      const [names, guessesBitfield] = content as unknown as [
        string[],
        Uint8Array,
      ];
      const guesses = bitfieldToBooleanArray(guessesBitfield, names.length);
      content = [names, guesses];
      break;
    }
    case MessageType.NameSubmitted: {
      const [name, id] = content as unknown as [string, Uint8Array];
      content = [name, stringify(id)];
      break;
    }
    case MessageType.UnsubmitName:
    case MessageType.NameUnsubmitted:
      content = stringify(content as unknown as Uint8Array);
      break;
  }

  return { type, content };
}

function booleanArrayToBitfield(array: boolean[]): Uint8Array {
  const bitfield = new Uint8Array(Math.ceil(array.length / 8));
  for (let i = 0; i < array.length; i++) {
    bitfield[Math.floor(i / 8)] |= Number(array[i]) << (7 - (i % 8));
  }
  return bitfield;
}

export function encodeMessage(message: Message): ArrayBuffer {
  let content: Uint8Array | null;
  switch (message.type) {
    case MessageType.Names:
      content = encode([
        message.content[0],
        booleanArrayToBitfield(message.content[1]),
      ]);
      break;
    case MessageType.NameSubmitted:
      content = encode([message.content[0], parse(message.content[1])]);
      break;
    case MessageType.UnsubmitName:
    case MessageType.NameUnsubmitted:
      content = encode(parse(message.content));
      break;
    case MessageType.StateSubmitting:
    case MessageType.SubmitName:
    case MessageType.NumNames:
    case MessageType.GuessName:
    case MessageType.NameGuessed:
    case MessageType.UnguessName:
    case MessageType.NameUnguessed:
      content = encode(message.content);
      break;
    case MessageType.RequestSubmittingState:
    case MessageType.RequestPlayingState:
      content = null;
      break;
  }

  const encoded = new ArrayBuffer(4 + (content?.length ?? 0));
  const view = new DataView(encoded);
  view.setUint32(0, message.type);

  if (content !== null) {
    new Uint8Array(encoded, 4).set(content);
  }

  return encoded;
}
